generator/src/test/java/com/scottlogic/deg/generator/cucumber/interesting/InSet.feature

```diff
@@ -0,0 +1,144 @@
+      | "foo" |
+      | "ABC" |
+      | "DEF" |
+
+Scenario: Running an 'inSet' request that includes 4 strings with alphabetic characters should return 2 values and null
+    Given there is a field foo
+    And foo is in set:
+      | "ABC" |
+      | "DEF" |
+      | "GHI" |
+      | "JKL" |
+    Then the following data should be generated:
+      | foo   |
+      | null  |
+      | "ABC" |
+      | "DEF" |
+
+Scenario: Running an 'inSet' request that includes only a single string with alphabetic characters should return 1 value and null
+    Given there is a field foo
+    And foo is in set:
+      | "Test" |
+    Then the following data should be generated:
+      | foo    |
+      | null   |
+      | "Test" |
+
+Scenario: Running an 'inSet' request that includes 4 numeric values and not null should return 2 values
+    Given there is a field foo
+    And foo is anything but null
+    And foo is in set:
+      | 10 |
+      | 20 |
+      | 30 |
+      | 40 |
+    Then the following data should be generated:
+      | foo |
+      | 20  |
+      | 40  |
+
+Scenario: Running an 'inSet' request that includes 4 numeric values should return 2 values and null
+    Given there is a field foo
+    And foo is in set:
+      | 100 |
+      | 2000 |
+      | 300 |
+      | 550 |
+    Then the following data should be generated:
+      | foo  |
+      | null |
+      | 100  |
+      | 2000 |
+
+ Scenario: Running an 'inSet' request that includes 4 decimal numeric values should return 2 decimal values and null
+    Given there is a field foo
+    And foo is in set:
+      | 100.05   |
+      | 253.0054 |
+      | 10.8     |
+      | 0.00382  |
+    Then the following data should be generated:
+      | foo      |
+      | null     |
+      | 10.8     |
+      | 253.0054 |
+
+Scenario: Running an 'inSet' request that includes 4 temporal values should return 2 temporal values and null
+    Given there is a field foo
+    And foo is in set:
+      | "2010-01-01T00:00:00.000" |
+      | "2012-01-01T00:01:00.000" |
+      | "2011-01-02T00:00:00.000" |
+      | "2012-01-02T01:10:00.000" |
+    Then the following data should be generated:
+      | foo                       |
+      | null                      |
+      | "2010-01-01T00:00:00.000" |
+      | "2012-01-01T00:01:00.000" |
+
+Scenario: Running an 'inSet' request that has multiple special characters should return 2 values and null
+    Given there is a field foo
+    And foo is in set:
+      | "Cyrillic text: Тхис ис Тест Нумбер 01" |
+      | "Japanese text: これはテスト番号2です"   |
+      | "Korean text: 이것은 시험 번호 3입니다"  |
+    Then the following data should be generated:
+      | foo |
+      | null |
+      | "Japanese text: これはテスト番号2です"   |
+      | "Korean text: 이것은 시험 번호 3입니다"  |
+
+Scenario: Running an 'inSet' request that has multiple set constraints should return 2 values from each set and null
+    Given there is a field foo
+    And there is a constraint:
+      """
+      {
+        "anyOf": [
```

Comment:
Hi, I was wondering how this would work with if constraints. Is this right? I expecting it to still return ABC and DEF (if constraint value + 1 other value) only but it also returned GHI. I might be missing something.


      Scenario: Running an 'inSet' request that includes 4 strings with alphabetic characters and an if constraint on one of the values should return the value in the if constraint plus one other value
          Given there is a field foo
          And foo is anything but null
          And foo is in set:
            | "ABC" |
            | "DEF" |
            | "GHI" |
            | "JKL" |
           And there is a constraint:
                """
                  {
                  "if": { "field": "foo", "is": "equalTo", "value": "ABC" },
                  "then": { "field": "foo", "is": "ofType", "value": "string" }
                  }
                """
          Then the following data should be generated:
            | "foo" |
            | "ABC" |
            | "DEF" |


Expected: [ABC], [DEF]
     but: [ABC], [DEF], [GHI]

Reply:
Essentially because of the conditional the set gets split before FieldSpecs are generated, so therefore you end up with a FieldSpec of [ABC] and a FieldSpec of [DEF, GHI, JKL] hence you end up with DEF and GHI being generated

Reply:
I think the above isn't wrong but agree that it does look like it could be.
I think its a bit of an edge case, maybe there is some additional merging/checking that can be performed to 'clean' up the emitted data.